Implemented flexberry-tabs

dpavloff opened the pull request:
- Added ability to scroll tabs with mousewheel or by dragging the tabs;
- In case tabs overflow its parent container, a dropdown appears, from which another tab can be chosen without scrolling;
- Implemented scrolling with mousewheel, ability to select new tab from dropdown;

KollegOFF reviewed (commented):
На самом деле есть необходимость реализовывать прокрутку на JS?
Можно же обычной браузерной прокруткой обойтись, она сейчас практически везде стилизуется, а где нет, там и не надо, мне кажется.

dpavloff commented:
Прокрутка была реализована таким образом, чтобы не требовалось зажимать shift при прокрутке. Решил, что это будет более удобным и интуитивным для пользователя решением.

KollegOFF commented:
Судя по описанию задачи из ТФС, зажимать `shift` как раз требуется, и мне кажется что, это поведение по умолчанию.
Нужно обсудить допустимость использования прокрутки, с её стилизацией и без.

lulzforme commented:
@KollegOFF когда писали постановку, то упор делался на ускорение разработки данного компонента. @dpavloff реализовал всё так как задумывается в макете, так что я не вижу проблем с прокруткой. Плюс такая прокрутка потом поможет нам с внедрением прокрутки в мобильную версию.

pepelyaeva commented:
https://github.com/Flexberry/ember-flexberry-themes/pull/89

dpavloff committed:
Implemented flexberry-tabs

Implemented flexberry-tabs

- Added ability to scroll tabs with mousewheel or by dragging the tabs;
- In case tabs overflow its parent container, a dropdown appears, from which another tab can be chosen without scrolling;
- Implemented scrolling with mousewheel, ability to select new tab from dropdown;

pepelyaeva commented on addon/styles/themes/default/modules/tab.overrides:
```diff
@@ -2,9 +2,73 @@
 *******************************/
 .flexberry-tab-bar {
-  float: right;
+  float: none;
```
Надо все стили перенести в blue-sky

Nasutya commented:
Созданный пример не открывается, в консоли ошибка:
TypeError: Cannot define property __ember_meta__, object is not extensible
    at Function.defineProperty (<anonymous>)
    at setMeta (ember.debug.js:18845)
    at Object.meta (ember.debug.js:18884)
    at addChainWatcher (ember.debug.js:14601)
    at new ChainNode (ember.debug.js:14649)
    at ChainNode.chain (ember.debug.js:14784)
    at ChainNode.chain (ember.debug.js:14793)
    at ChainNode.add (ember.debug.js:14745)
    at Object.watchPath (ember.debug.js:23562)
    at Object.watch (ember.debug.js:23614)

TeterinaSvetlana committed:
Merge branch 'develop' into feature-203161

TeterinaSvetlana committed:
code review

pepelyaeva commented:
Нужно исправить под 3 версию

eneroslov committed:
Merge branch 'develop' into feature-203161

eneroslov committed:
fix bugs

eneroslov committed:
created tests

eneroslov committed:
fix tests

eneroslov commented:
ПР с документацией по flexberry-tab-bar: https://github.com/Flexberry/flexberry.github.io/pull/164

DubrovinPavel commented on tests/integration/components/flexberry-tab-bar-test.js:
```diff
@@ -0,0 +1,247 @@
+  assert.strictEqual($dropdown.hasClass('dropdown link item'), true, 'Component\'s wrapper has \'dropdown icon\' css-class');
+});
+
+test('it renders items', function(assert) {
```
А не получится этот тест объединить с 2 следующими? У них вроде рендерится одно и то же и в результате много повторения

eneroslov replied:
Объединил 2 последних теста с переходам по вкладкам через меню и выпадающий список.

pepelyaeva commented on tests/dummy/app/routes/components-examples/flexberry-tab-bar/settings-example.js:
```diff
@@ -0,0 +1,4 @@
+import Ember from 'ember';
+
+export default Ember.Route.extend({
+});
```
import Route from '@ember/routing/route';

export default Route.extend({
});

pepelyaeva commented on tests/dummy/app/controllers/components-examples/flexberry-tab-bar/settings-example.js:
```diff
@@ -0,0 +1,54 @@
+import Ember from 'ember';
```
import Controller from '@ember/controller';

pepelyaeva commented on tests/dummy/app/controllers/components-examples/flexberry-tab-bar/settings-example.js:
```diff
@@ -0,0 +1,54 @@
+import { htmlSafe } from '@ember/string';
+import { computed } from '@ember/object';
+
+export default Ember.Controller.extend({
```
export default Controller.extend({

pepelyaeva commented on addon/styles/themes/blue-sky/modules/tab.overrides:
Этот файл нужно удалить

pepelyaeva commented on addon/styles/themes/blue-sky/modules/tab.variables:
Этот файл нужно удалить

eneroslov committed:
fix lint

eneroslov committed:
fix tests

pepelyaeva committed:
Add isOverflowedTabs setting

pepelyaeva committed:
Merge branch 'develop' into feature-203161

pepelyaeva committed:
Up ef-themes version

pepelyaeva committed:
Fix tests

DubrovinPavel commented on addon/components/flexberry-tab-bar.js:
```diff
@@ -48,12 +51,155 @@ export default Component.extend({
+   * @method normalizeWheel
+   */
+  normalizeWheel: function (event) {
+    var pixelStep = 10;
+    var lineHeight = 40;
+    var pageHeight = 800;
+    var sX = 0;
+    var sY = 0;
+    var pX = 0;
+    var pY = 0;
```
Заменить на let/const

DubrovinPavel commented on addon/components/flexberry-tab-bar.js:
```diff
@@ -124,6 +247,68 @@ export default Component.extend({
+        this.$('body').stop();
+
+        // Calculate delta, direction.
+        var n = this.normalizeWheel(e);
+        var x = n.pixelX !== 0 ? n.pixelX : n.pixelY;
+        var delta = Math.min(Math.abs(x), 150);
+        var direction = x > 0 ? 1 : -1;
```
Заменить на let/const

DubrovinPavel commented on tests/integration/components/flexberry-tab-bar-test.js:
```diff
@@ -0,0 +1,162 @@
+    let $item = $(this);
+
+    // Check that the captions matches the array.
+    assert.strictEqual($item.attr('data-tab'), 'tab' + new Number(i + 1), 'Component\'s item\'s сaptions matches the array');
+  });
+
+  $dropdownItems.each(function(i) {
+    let $item = $(this);
+
+    // Check that the captions matches the array.
+    assert.strictEqual($item.attr('data-tab'), 'tab' + new Number(i + 1), 'Component\'s item\'s сaptions matches the array');
```
Sonar на эти Number еще ругается, я с ним согласен в принципе. Можно что-то вроде `tab${i + 1}` сделать

eneroslov committed:
Fix remarks
